BlockHeaderRegistry

pmlambert opened the pull request.

pmlambert commented:
@leonprou I believe I'm supposed to sync up with you to figure out exactly how to structure the runtime logic for the changes. Everything works here, I just need to know exactly where you guys want to go with this feature. For example, who can submit signatures for which blocks (are cycles important), can validators submit signatures for multiple blocks at the same height, how many blocks should be confirmed before submitting the signatures, etc.

sirpy commented on app/index.js:
```diff
@@ -34,6 +37,37 @@ function initWalletProvider() {
+    if (chainId == 122) {
+      let cycleEnd = (await consensus.methods.getCurrentCycleEndBlock.call()).toNumber()
+      let numValidators = (await consensus.methods.currentValidatorsLength.call()).toNumber()
+      const validators = await Promise.all(new Array(numValidators).map(async (_, i) => await consensus.methods.currentValidatorsAtPosition.call()))
```
better use multicall here to fetch cycleend,numValidators,validators

sirpy commented on app/index.js:
```diff
@@ -34,6 +37,37 @@ function initWalletProvider() {
+    signer: new ethers.Wallet(pkey)
+    blocks: {},
+  }
+  blockchains[chainId].web3.eth.subscribe('newBlockHeaders', async (block) => {
```
you are not handling exceptions

sirpy commented on app/index.js:
```diff
@@ -120,6 +158,35 @@ function emitRewardedOnCycle() {
 
+async function emitRegistry() {
+  logger.info('emitRegistry')
+  const chains = await blockRegistry.getPastEvents('Blockchain', {fromBlock:0,toBlock:'latest'})
+  await Promise.all(chains.filter(chain => blockchains[chain[0]].rpc != chain[1] || !blockchains[chain[0]]).map(async (chain) => initBlockchain(...chain)))
```
`blockchains[chain[0]].rpc != chain[1]`
this will fail with undefined
`!blockchains[chain[0]]` needs to be the first condition

sirpy commented on scripts/quickstart.sh:
```diff
@@ -43,6 +43,8 @@ DEFAULT_GAS_ORACLE="https:\/\/ethgasstation.info\/json\/ethgasAPI.json"
 
 PARITY_SNAPSHOT="https://node-snapshot.s3.eu-central-1.amazonaws.com/db.tar.gz"
 OE_SNAPSHOT="https://node-snapshot-oe.s3.eu-central-1.amazonaws.com/db.tar.gz"
+BOOTNODE_FILE_FUSE="https://raw.githubusercontent.com/fuseio/fuse-network/master/config/bootnodes.txt"
+BOOTNODE_FILE_SPARK="https://raw.githubusercontent.com/fuseio/fuse-network/master/config/spark/bootnodes.txt"
```
what are all these changes to the quickstart script?

sirpy commented on Dockerfile.spark:
```diff
@@ -1,5 +1,5 @@
-FROM alpine:3.13.2 AS builder
-  
+FROM alpine:edge AS builder
```
are all these changes approved by fuse?

sirpy reviewed (changes requested):
1. contracts should not be here, they are irrelevant to the fuseapp
2. what are all the changes to fuseapp files not related to the bridge?
3. you are missing unit tests for the fuseapp bridge part, it is part of the bounty requirements
4. `By default fuse+ethereum+bsc are always enabled no matter if they are in the list or not`

sirpy commented:
> * contracts should not be here, they are irrelevant to the fuseapp
> * what are all the changes to fuseapp files not related to the bridge?
> * you are missing unit tests for the fuseapp bridge part, it is part of the bounty requirements
> * `By default fuse+ethereum+bsc are always enabled no matter if they are in the list or not`

@pmlambert you missed this comment

You are still not handling errors correctly in the .subscribe, better extract the whole .subscribe callback to a function

sirpy commented on app/index.js:
```diff
@@ -120,6 +152,40 @@ function emitRewardedOnCycle() {
+    logger.info('emitRegistry')
+    const currentBlock = (await web3.eth.getBlockNumber()).toNumber()
+    const chains = await blockRegistry.methods.getRpcs.call()
+    await Promise.all(chains.filter(chain => !blockchains[chain[0]] || blockchains[chain[0]].rpc != chain[1]).map(async (chain) => initBlockchain(...chain)))
```
if one initblockchain fail it shouldnt fail the whole function

sirpy replied:
still not done

sirpy commented on app/index.js:
```diff
@@ -34,6 +37,31 @@ function initWalletProvider() {
+      blocks: {},
+    }
+    blockchains[chainId].web3.eth.subscribe('newBlockHeaders', async (block) => {
+      if (chainId == 122) {
+        let cycleEnd = (await consensus.methods.getCurrentCycleEndBlock.call()).toNumber()
+        let validators = await consensus.methods.currentValidators().call()
+        const numValidators = validators.length
+        blockchains[chainId].blocks[block.hash] = await signFuse(block, chainId, blockchain.provider, blockchain.signer, cycleEnd, validators)
+      }
+      else {
+        blockchains[chainId].blocks[block.hash] = await sign(block, chainId, blockchain.provider, blockchain.signer)
+      }
+    })
```
handle exceptions and extract to a function

pmlambert commented:
> 1. contracts should not be here, they are irrelevant to the fuseapp
Deleted
> 2. what are all the changes to fuseapp files not related to the bridge?
From using rebase, has been fixed
> 3. you are missing unit tests for the fuseapp bridge part, it is part of the bounty requirements
That is step 3, this is for parts 1 and 2
> 4. `By default fuse+ethereum+bsc are always enabled no matter if they are in the list or not`
Changed to use env keys ETH,BSC _RPC

sirpy commented:
> > 1. contracts should not be here, they are irrelevant to the fuseapp
> >    Deleted
> > 2. what are all the changes to fuseapp files not related to the bridge?
> >    From using rebase, has been fixed
> > 3. you are missing unit tests for the fuseapp bridge part, it is part of the bounty requirements
> >    That is step 3, this is for parts 1 and 2
> > 4. `By default fuse+ethereum+bsc are always enabled no matter if they are in the list or not`
> >    Changed to use env keys ETH,BSC _RPC

Every step has its own requirement for unit tests. you can check the bounty and check boxes again for each step.

sirpy commented on app/index.js:
```diff
@@ -55,6 +89,13 @@ function initBlockRewardContract() {
+  logger.info(`initBlockRegistryContract`, process.env.BLOCK_REGISTRY_ADDRESS)
+  blockRegistry = new web3.eth.Contract(require(path.join(cwd, 'abi/blockRegistry')), process.env.BLOCK_REGISTRY_ADDRESS)
+  initBlockchain(1, process.env.ETH_RPC || throw "Missing ETH_RPC in environment")
+  initBlockchain(56, process.env.BSC_RPC || throw "Missing BSC_RPC in environment"))
```
bsc has public rpc

sirpy commented on app/index.js:
```diff
@@ -55,6 +89,13 @@ function initBlockRewardContract() {
+function initBlockRegistryContract() {
+  logger.info(`initBlockRegistryContract`, process.env.BLOCK_REGISTRY_ADDRESS)
+  blockRegistry = new web3.eth.Contract(require(path.join(cwd, 'abi/blockRegistry')), process.env.BLOCK_REGISTRY_ADDRESS)
+  initBlockchain(1, process.env.ETH_RPC || throw "Missing ETH_RPC in environment")
```
use cloudflare eth endpoint

sirpy commented on app/index.js:
```diff
@@ -120,6 +161,40 @@ function emitRewardedOnCycle() {
+      })
+    })
+  } catch(e) {
+    throw `emitRegistry failed trying to update rpcs`
```
i dont think that should prevent from calling addSignedBlocks part

sirpy commented on app/index.js:
```diff
@@ -132,13 +207,17 @@ async function runMain() {
     }
     await emitInitiateChange()
     await emitRewardedOnCycle()
+    await emitRegistry()
```
i dont think an exception here should cause the fuseapp to exit

sirpy commented on app/index.js:
```diff
@@ -120,6 +161,40 @@ function emitRewardedOnCycle() {
+  try {
+    logger.info('emitRegistry')
+    const currentBlock = (await web3.eth.getBlockNumber()).toNumber()
+    const chains = await blockRegistry.methods.getRpcs.call()
```
should be getRpcs()
and that's why I want unit tests or testnet deployment with mock contracts to see that this whole process is working
